src/librustc_typeck/collect.rs

```diff
@@ -50,9 +49,65 @@ use rustc::hir::GenericParamKind;
 use rustc::hir::{self, CodegenFnAttrFlags, CodegenFnAttrs, Unsafety};
 
 use std::iter;
+use std::ops::Range;
 
 struct OnlySelfBounds(bool);
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LintedBoundVar<'tcx> {
+    DefId(DefId),
+    SelfTy,
+    Ty(Binder<Ty<'tcx>>),
+    Region(Binder<ty::Region<'tcx>>),
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LintedBoundValue<'tcx> {
+    DefId(DefId),
+    Ty(Binder<Ty<'tcx>>),
+    Region(Binder<ty::Region<'tcx>>),
+}
+
+pub fn lint_duplicate_bounds<'a, 'gcx, 'tcx>(
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    node_id: ast::NodeId,
+    bounds: impl IntoIterator<Item = (LintedBoundVar<'tcx>, LintedBoundValue<'tcx>, Span)>,
+) {
+    let mut bounds: Vec<_> = bounds.into_iter().collect();
+    bounds.sort_unstable();
+
+    let emit_lint = |range: Range<usize>| {
+        let bound_name = tcx.sess.source_map()
+                            .span_to_snippet(bounds[range.start].2).unwrap();
+        let mut err = tcx.struct_span_lint_node(
+            lint::builtin::DUPLICATE_BOUNDS,
+            node_id,
+            bounds[range.clone()].iter().map(|(_, _, sp)| *sp).collect::<Vec<_>>(),
+            &format!("duplicate bound `{}` found",
+                     bound_name));
+        debug!("zzz: 1: {:?} / {:?}", bounds[range.start].0, bounds[range.start].1);
+        err.span_label(bounds[range.start].2, "first use of bound");
+        for i in (range.start + 1)..range.end {
```

Comment:
Why not get a subslice and iterate over that + pattern match out the parts of the tuple?

Reply:
Because it's important to have the indices, as you can see from the following code.

Reply:
You can use `.enumerate()` to get the indices :)

Reply:
Yeah, but the thing is I'm keeping track of the start index of a "run"... this sort of made most sense to me. *shrug*